### ISSUE TITLE
How are you maintaining redux state over page refreshes?

### ISSUE THREAD

**Issue by roblapp** [maxmantz/redux-oidc#37]
I have integrated my app to work with React-Redux-Oidc. However, one issue I am having is that when I have a user who is logged in and I refresh the page the state.oidc.user is null again. Thus, my app thinks the user is logged out even though I still see the OIDC metadata in session storage.

When I ran your example application locally I see that you did not have this issue. But I am unable to figure out what mechanism on the example app is keeping state.oidc.user persistent on page refreshes?

**Comment by maxmantz** [comment]
The middleware loads the user data from session storage into the reducer when the app first loads. Please make sure your middleware is set up correctly.

**Comment by roblapp** [comment]
Whenever I refresh I lose the user in the redux state. I can't figure out why. Here is my code:

```
//index.js
import 'babel-polyfill'; //MUST HAVE THIS HERE!!!
import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { OidcProvider } from 'redux-oidc';
import configureStore from './store/configureStore';
import AppContainer from './components/App';
import userManager from './config/UserManager';
import './index.css';

const store = configureStore();

ReactDOM.render(
    <Provider store={store}>
      <OidcProvider store={store} userManager={userManager}>
        <BrowserRouter>
            <App store={store} userManager={userManager} />
        </BrowserRouter>
      </OidcProvider>
    </Provider>,
  document.getElementById('root')
);

//reducers/index.js
import { combineReducers } from 'redux';
import todoReducer from './todoReducer';
import { reducer as oidcReducer } from 'redux-oidc';

const rootReducer = combineReducers({
    todos: todoReducer,
    oidc: oidcReducer
});

export default rootReducer;


//configureStore.js
import {createStore, compose, applyMiddleware} from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import rootReducer from '../reducers';
import createOidcMiddleware from 'redux-oidc';
import userManager from '../config/UserManager';

const oidcMiddleware = createOidcMiddleware(userManager);

function configureStore(initialState) {
  const middlewares = [
    // Add other middleware on this line...

    oidcMiddleware,

    // thunk middleware can also accept an extra argument to be passed to each thunk action
    // https://github.com/gaearon/redux-thunk#injecting-a-custom-argument
    thunk,
  ];

  const store = createStore(rootReducer, initialState, compose(applyMiddleware(...middlewares)));

  return store;
}

export default configureStore;

//UserManager.js
import { createUserManager } from 'redux-oidc';
import { Log } from 'oidc-client';

const userManagerConfig = {
    authority: 'https://demo.identityserver.io',
    client_id: 'implicit',
    redirect_uri: 'http://localhost:3000/Callback',
    response_type: 'id_token token',
    scope: 'openid profile email api',
    post_logout_redirect_uri: 'http://localhost:3000'
};

Log.logger = console;

const userManager = createUserManager(userManagerConfig);

export default userManager;

//App.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import AuthenticatedApp from './AuthenticatedApp';
import DefaultApp from './DefaultApp';
import CallbackContainer from '../../../components/Callback';
import 'bootstrap/dist/css/bootstrap.css';
import 'bootstrap/dist/css/bootstrap-theme.css';
import '../styles/App.css';

class App extends Component {

  render() {
    console.log("App::render()");
    console.log(this.props.store.getState().oidc.user);

    const oidcState = this.props.store.getState().oidc;
    if (oidcState && oidcState !== null && oidcState.user) {
      return (
        <AuthenticatedApp userManager={this.props.userManager} />
      );
    }

    return (
      <DefaultApp userManager={this.props.userManager} />
    );
  }
}

App.propTypes = {
    userManager: PropTypes.object.isRequired,
    store: PropTypes.object.isRequired
};

export default App;

```

And my Callback is exactly as you have in your sample app. Everything works fine except on refresh when I lose the user info from the state... although it stays in sessionStorage... Any suggestions?

**Comment by maxmantz** [comment]
I've received an email with a comment from you that you subsequently deleted. Were you able to solve your issue?

**Comment by roblapp** [comment]
Yes, I believe I know what the issue is now. When I have a concise answer I will post it here. Thanks!

**Comment by maxmantz** [comment]
Good to hear. I'd be happy to see your answer.

**Comment by roblapp** [comment]
While I haven't figured out why your example app works and mine doesn't I can explain what I have found so far.

When the application refreshes the middlewareHandler generator function (from redux-oidc) is not running to trigger the user to be loaded into the redux state. Although oidc-js does load the user which I can see through logging, they are not firing an event for this. It appeas that when a UserManager object gets created it triggers the getUser function seen here [https://github.com/IdentityModel/oidc-client-js/blob/dev/src/UserManager.js#L57](url). Notice on line 64 the function is calling this._events.load(user, false);. This load function is implemented here [https://github.com/IdentityModel/oidc-client-js/blob/dev/src/UserManagerEvents.js#L19](url). You'll notice that the second parameter in the load() is being passed in as false. Therefore, it is bypassing the userLoad event to be raised by oidc-js. This sucks because you have an event handler for this event setup in the OidcProvider.js file. Specifically, the function here [https://github.com/maxmantz/redux-oidc/blob/master/src/OidcProvider.js#L36](url) handles the userFound event when it's fired, which then invokes the middlewareHandler generator function here [https://github.com/maxmantz/redux-oidc/blob/master/src/oidcMiddleware.js#L19](url).

So to summarize, the event that triggers the middleware to run is not being fired. So I suppose I need to figure out what I am doing wrong. To be clear I have everything working except when the application is re-loaded. I haven't been able to figure out why it's working in the example app and not my app.

Does this make sense?

**Comment by maxmantz** [comment]
You're not doing anything "wrong" in particular. The middleware is triggered by an action. You could use some action like this after creating the redux store:

```
store.dispatch({type: 'INIT_APP'});
```
This should kick the middleware into gear and start the validation process.

**Comment by maxmantz** [comment]
Were you able to solve this?

**Comment by roblapp** [comment]
Sorry for the delayed response. Yes, I was able to solve this by firing the event you mentioned.  Obviously this dispatched action triggers the middleware and all works as expected. Thank you for your correspondence!

**Comment by acaravia** [comment]
I know this thread is old but I'm experiencing this same issue with the latest version which doesn't use the middleware anymore so wondering if there is another solution? Everything works fine, slient renew works as long as you don't do a browser refresh. Once that occurs it appears that silent renew no longer works. I'm using react-router-redux as well.....I can't seem to pinpoint the issue but any thoughts would be appreciated. I'll return with more information once I have more.....thanks
